mypy_django_plugin/transformers/models.py

```diff
@@ -640,19 +641,41 @@ def get_exception_bases(self, name: str) -> List[Instance]:
 
         return bases
 
-    def add_exception_classes(self, model_cls: Type[Model]) -> None:
+    @cached_property
+    def is_model_abstract(self) -> bool:
+        meta = self.model_classdef.info.names.get("Meta")
+        # Check if 'abstract' is declared in this model's 'class Meta' as
+        # 'abstract = True' won't be inherited from a parent model.
+        if meta is not None and isinstance(meta.node, TypeInfo) and "abstract" in meta.node.names:
+            abstract = meta.node.names["abstract"].node
+            for stmt in meta.node.defn.defs.body:
+                if (
+                    # abstract =
+                    isinstance(stmt, AssignmentStmt)
+                    and len(stmt.lvalues) == 1
+                    and isinstance(stmt.lvalues[0], NameExpr)
+                    and stmt.lvalues[0].name == "abstract"
+                    # = True (builtins.bool)
+                    and isinstance(stmt.rvalue, NameExpr)
+                    and stmt.rvalue.name == "True"
+                    and isinstance(abstract, Var)
+                    and isinstance(abstract.type, Instance)
+                    and abstract.type.type.fullname == "builtins.bool"
+                ):
+                    return True
+        return False
+
+    def add_exception_classes(self) -> None:
         """
         Adds exception classes 'DoesNotExist' and 'MultipleObjectsReturned' to a model
         type, aligned with how the model metaclass does it runtime.
 
         If the model is abstract, exceptions will be added as abstract attributes.
         """
-        is_abstract = model_cls._meta.abstract
```

Comment:
Can you please explain why do you prefer static check over runtime one?

Reply:
Runtime one didn't work. As abstract models aren't returned by `apps.get_models()`, we just happen to append any abstract model bases we run in to when looking through `apps.get_models()`.

But that becomes an issue with e.g. `AbstractUser` as it is then dependant on having a concrete model (i.e. in `apps.get_models()`) that has subclassed `AbstractUser` to be able to process `AbstractUser`.

Ref:

https://github.com/typeddjango/django-stubs/blob/a8e42cbcca7a2ce3d890789223d3043412e51c4c/mypy_django_plugin/django/context.py#L88-L98

Reply:
Hm, maybe we should modify this plugin hook: https://github.com/typeddjango/django-stubs/blob/a8e42cbcca7a2ce3d890789223d3043412e51c4c/mypy_django_plugin/main.py#L259-L263 to work with all classes that have `Model` base class in its `mro`?

This way we can still use the runtime API which is the most reliable way to analyze things.

Reply:
I think it's shorter/better to look for `isinstance(type_info.metaclass_type, Instance) and type_info.metaclass_type.type.fullname == "django.db.models.base.ModelBase"` to figure out what is a model class.

But changing the condition you mention won't affect if we find a model or not. We won't be given all abstract models from the runtime API.

Reply:
Changing the hook condition seems to propagate into removing some more stuff. Perhaps we could do that in a separate PR?

Reply:
I've opened #1668 just now